community/kernel/src/main/java/org/neo4j/kernel/impl/storemigration/StoreUpgrader.java

```diff
@@ -150,32 +151,42 @@ public void migrateIfNeeded( File storeDirectory, SchemaIndexProvider schemaInde
             throw e;
         }
 
-        try
+        MigrationStatus migrationStatus = readMigrationStatus( migrationStateFile );
+        String versionToMigrateFrom = null;
+        // We don't need to migrate if we're at the phase where we have migrated successfully
+        // and it's just a matter of moving over the files to the storeDir.
+        if ( migrationStatus == null || migrationStatus.ordinal() <= MigrationStatus.migrating.ordinal() )
         {
-            File migrationStateFile = new File( migrationDirectory, MIGRATION_STATUS_FILE );
+            versionToMigrateFrom = upgradableDatabase.checkUpgradeable( storeDirectory );
+            cleanMigrationDirectory( migrationDirectory );
+            setMigrationStatus( migrationStateFile, MigrationStatus.migrating, null );
+            migrateToIsolatedDirectory( storeDirectory, migrationDirectory, schemaIndexProvider, versionToMigrateFrom );
+            setMigrationStatus( migrationStateFile, MigrationStatus.moving, versionToMigrateFrom );
+        }
 
-            // We don't need to migrate if we're at the phase where we have migrated successfully
-            // and it's just a matter of moving over the files to the storeDir.
-            if ( !migrationStatusIs( migrationStateFile, MigrationStatus.moving ) )
+        if ( migrationStatus == null || migrationStatus.ordinal() <= MigrationStatus.moving.ordinal() )
```

Comment:
Handling of null values in `migrationStatus` and `versionToMigrateFrom` is overcomplicated and tricky.
WDYT about updating migrationStatus to current status in each step instead and just read `versionToMigrateFrom` since that will be always read except we are in `completed` phase that should not be the case here